httpbp/middlewares.go

```diff
@@ -404,3 +417,102 @@ func recordStatusCode(counters counterGenerator) Middleware {
+//
+//   - http_method: method of the HTTP request
+//   - http_endpoint: path to identify the endpoint handler, may be empty
+//
+// * http_server_latency_seconds, http_server_request_size_bytes, http_server_response_size_bytes histograms with labels above plus:
+//
+//   - http_success: true if the status code is 2xx or 3xx, false otherwise
+//
+// * http_server_requests_total counter with all labels above plus:
+//
+//   - http_response_code: numeric status code as a string, e.g. 200
+func PrometheusServerMetrics(serverSlug string) Middleware {
+	return func(name string, next HandlerFunc) HandlerFunc {
+		return func(ctx context.Context, w http.ResponseWriter, r *http.Request) (err error) {
+			start := time.Now()
+			method := r.Method
+			endpoint := r.URL.Path
+			activeRequestLabels := prometheus.Labels{
+				methodLabel:   method,
+				endpointLabel: endpoint,
+			}
+			serverActiveRequests.With(activeRequestLabels).Inc()
+
+			wrapped := &responseRecorder{ResponseWriter: w}
+			defer func() {
+				code := processCodeErr(err, wrapped.responseCode)
+				success := isRequestSuccessful(code, err)
+
+				labels := prometheus.Labels{
+					methodLabel:   method,
+					successLabel:  success,
+					endpointLabel: endpoint,
+				}
+				serverLatency.With(labels).Observe(time.Since(start).Seconds())
+				serverRequestSize.With(labels).Observe(float64(r.ContentLength))
+				serverResponseSize.With(labels).Observe(float64(getContentLength(w)))
+
+				totalRequestLabels := prometheus.Labels{
+					methodLabel:   method,
+					successLabel:  success,
+					endpointLabel: endpoint,
+					codeLabel:     strconv.Itoa(code),
+				}
+				serverTotalRequests.With(totalRequestLabels).Inc()
+				serverActiveRequests.With(activeRequestLabels).Dec()
+			}()
+
+			return next(ctx, wrapped, r)
+		}
+	}
+}
+
+// isRequestSuccessful returns the success of an HTTP request as a string, i.e. "true" or "false".
+// A HTTP request is successful when:
+//   1) no error is returned from the request and
+//   2) the HTTP status code is in the form 2xx.
+func isRequestSuccessful(httpStatusCode int, requestErr error) string {
+	return strconv.FormatBool(requestErr == nil && isSuccessStatusCode(httpStatusCode))
+}
+
+const contentLengthHeader = "Content-Length"
+
+func getContentLength(w http.ResponseWriter) int64 {
+	if cl := w.Header().Get(contentLengthHeader); cl != "" {
+		v, err := strconv.ParseInt(cl, 10, 64)
+		if err == nil && v >= 0 {
+			return v
+		}
+	}
+	return 0
+}
+
+// responeRecorder records the following:
+//   1. HTTP response code passed to a call to WriteHeader.
+//   2. the content length of whats written.
+type responseRecorder struct {
+	http.ResponseWriter
+
+	bytesWritten int
+	responseCode int
+}
+
+func (rr *responseRecorder) Write(b []byte) (n int, err error) {
+	n, err = rr.ResponseWriter.Write(b)
+	rr.bytesWritten += n
+	rr.ResponseWriter.Header().Set(contentLengthHeader, strconv.Itoa(rr.bytesWritten))
```

Comment:
I don't think we want to do this? this actually changes the behavior from what the underlying `http.ResponseWriter.Write` does.

Reply:
I must have misunderstood...i was trying to implement "I think we should wrap ResponseWriter to actually count the bytes written" from this comment https://github.com/reddit/baseplate.go/pull/451#discussion_r774737561

Reply:
the part we need to do here is only to count the bytes, nothing extra.

if `ResponseWriter.Write(b)` writes the content length header, then it's already taken care of on 509 and you do not need to do it again here. if `ResponseWriter.Write(b)` does not write the content length header, then doing it here changes that behavior.

Reply:
and I'm like 80% sure when using `ResponseWriter.Write` go actually uses chunked responses without content-length header.

Reply:
i see. makes sense. thanks. i will fix.